kitsune/messages/utils.py

```diff
@@ -1,39 +1,56 @@
+from django.contrib.auth.models import Group
+
 from kitsune.messages.models import InboxMessage, OutboxMessage
 from kitsune.users.models import User
 from kitsune.messages.signals import message_sent
 from kitsune.messages.tasks import email_private_message
 from kitsune.users.models import Setting
 
 
-def send_message(to, to_group=None, text=None, sender=None):
+def send_message(to, text=None, sender=None):
     """Send a private message.
 
-    :arg to: a list of Users to send the message to
-    :arg to_groups: Groups to send the message to
+    :arg to: Users or Groups to send the message to
     :arg sender: the User who is sending the message
     :arg text: the message text
     """
-    if not sender or not text:
-        return
 
-    # Assuming OutboxMessage doesn't necessarily need to be created for every recipient
-    if to:
-        outbox_message = OutboxMessage.objects.create(sender=sender, message=text)
-        outbox_message.to.set(to)
-        if to_group:
-            outbox_message.to_group.set(to_group)
-
-    for recipient in to:
-        if isinstance(recipient, User):
-            inbox_message = InboxMessage.objects.create(sender=sender, to=recipient, message=text)
-            # If we had a user, and we made them an inbox message,
-            # we should also add the to_groups to their message as well
-            if to_group:
-                inbox_message.to_group.set(to_group)
-            if Setting.get_for_user(recipient, "email_private_messages"):
-                email_private_message(inbox_message_id=inbox_message.id)
-
-    message_sent.send(sender=InboxMessage, to=to, to_group=to_group, text=text, msg_sender=sender)
+    # We need a sender, a message, and someone to send it to
+    if not sender or not text or not to:
+        return
+    # This is users from the To field
+    users = {user for user in to if isinstance(user, User)}
+    # This is groups from the To field
+    groups = {group for group in to if isinstance(group, Group)}
+    # This is all the users that are going to receive the message
+    receivers = users | set(User.objects.filter(groups__in=groups).distinct())
+
+    outbox_message = OutboxMessage.objects.create(sender=sender, message=text)
+    # Add the users from the To field to the outbox message, not all the users
+    # that are going to receive the message - this way we don't overwhelm the
+    # message UI
+    outbox_message.to.set(users)
+
+    if groups:
+        # Add the groups from the To field to the message
+        outbox_message.to_group.set(groups)
+
+    set_of_user_pks_to_email_private_message = set(
+        Setting.objects.filter(
+            user__in=receivers, name="email_private_messages", value=True
+        ).values_list("user__pk", flat=True)
```

Comment:
Just for the sake of knowledge sharing, `.values_list("user", flat=True)` is equivalent to `.values_list("user__pk", flat=True)`.

Reply:
Ah - yes, I left that when troubleshooting why this was failing. It was the `value=True` bit that was missing.

Reply:
Good catch on the `value=True`!